### ISSUE TITLE
Regression in testextra/grpauto.tst (memory usage)

### ISSUE THREAD

**Issue by alex-konovalov** [gap-system/gap#2377]
The following is observed in `teststandard` tests on Jenkins:
```
########> Diff in /circa/scratch/gap-jenkins/workspace/GAP-master-test/GAPCOPT\
S/64build/GAPTARGET/standard/label/kovacs/GAP-master-snapshot/tst/testextra/gr\
pauto.tst:105
# Input is:
IsomorphismGroups(G,PcGroupCode(CodePcGroup(G),Size(G)))=fail;
# Expected output:
false
# But found:
Error, reached the pre-set memory limit
(change it with the -o command line option)
########
```
In the master branch this appens on both macos and linux jenkins slaves. In the stable-4.9 branch - only on macos. I've tried to run just that example starting GAP with `-o 1g` option and so war was able to reproduce this problem only once. Apparently, there is some interaction between other tests and loaded packages, and balancing around the 1g limit with some randomness and garbage collections making this hard to reproduce.

This is the input:
```
LoadAllPackages();
G:=PcGroupCode( 741231213963541373679312045151639276850536621925972119311,11664);;
H:=PcGroupCode( 888658311993669104086576972570546890038187728096037768975,11664);;
IsomorphismGroups(G,H);
IsomorphismGroups(G,PcGroupCode(CodePcGroup(G),Size(G)))=fail;
```
and this is the memory usage that I observe under macos:
```
Load GAP                131           131 
Load all packages       ---           344
After 1st iso test      142           436
After 2nd iso test      548           1.01g
```
(the last 1.01g value is measured after a break loop happened and I've entered "return" to resume the calculation).

The problem can be reproduced easier if GAP is started with just slightly smaller memory limit `-o 1000m`. Then after `Error, reached the pre-set memory limit` I see this:

```
brk> Where(50);
ImagesRepresentative( map, gen ) at /Users/alexk/GITREPS/gap-stable/lib/mapphomo.gi:311 called from
func( C[i] ) at /Users/alexk/GITREPS/gap-stable/lib/coll.gi:746 called from
List( GeneratorsOfMagmaWithInverses( elms ), function ( gen )
      return ImagesRepresentative( map, gen );
  end ) at /Users/alexk/GITREPS/gap-stable/lib/mapphomo.gi:311 called from
ImagesSet( map, elm ) at /Users/alexk/GITREPS/gap-stable/lib/mapping.gi:185 called from
Image( hom, sub ) at /Users/alexk/GITREPS/gap-stable/lib/autsr.gi:811 called from
act( orb[p], acts[i] ) at /Users/alexk/GITREPS/gap-stable/lib/oprt.gi:1241 called from
OrbitStabilizerAlgorithm( G, false, false, gens, acts, rec(
    pnt := d,
    act := act,
    onlystab := true ) ) at /Users/alexk/GITREPS/gap-stable/lib/oprt.gi:2909 called from
orbish( G, pnt, gens, acts, act ) at /Users/alexk/GITREPS/gap-stable/lib/oprt.gd:866 called from
CallFuncList( StabilizerFunc, arg ) at /Users/alexk/GITREPS/gap-stable/pkg/rcwa-4.6.1/lib/rcwagrp.gi:4586 called from
Stabilizer( api, v, gens, pre, asAutomorphism ) at /Users/alexk/GITREPS/gap-stable/lib/autsr.gi:953 called from
PatheticIsomorphism( G, H ) at /Users/alexk/GITREPS/gap-stable/lib/morpheus.gi:2101 called from
<function "IsomorphismGroups">( <arguments> )
 called from read-eval loop at *errin*:1
brk> arg[1];
[ f4, f3*f4, f5, f6, f8^2*f10^2*f11, f8*f9^2*f12*f13*f14*f19^2, f13*f14, f13, f16, f15*f16, f8^2*f10*f17*f19^2*f20^2, f18, 
  f19, f20 ]
brk> arg[2];
function( gen ) ... end
brk> Print(last);
function ( gen )
    return ImagesRepresentative( map, gen );
end
brk>
```

**Comment by fingolfin** [comment]
Woah, not so fast. This is the original `Stabilizer` function from the library:
```gap
InstallGlobalFunction( Stabilizer, function( arg )
    if Length( arg ) = 1  then
        return StabilizerOfExternalSet( arg[ 1 ] );
    else
        return CallFuncList( StabilizerFunc, arg );
    fi;
end );
```
So while the above hack is, well, a hack, I don't quite see how it would cause the problem at hand, given that we correctly branch into `StabilizerFunc`

**Comment by alex-konovalov** [comment]
Yes, just came to the same conclusion looking at the same code of `Stabilizer`.

**Comment by alex-konovalov** [comment]
Looking now at some packages that appear in the results of profiling and are not loaded by default. That's what I've got so far:

| packages   | Load GAP | load packages | 1st iso test | 2nd iso test
| --- | --- | --- | --- | --- |
default      |    131   |       ---     |        142   |         548    |
all pkg      |    131   |       344     |        436   |         1.01g  |
FR           |    131   |       131     |        169   |         596    |
semigroups   |    131   |       131     |        193   |   720            |

**Comment by fingolfin** [comment]
I wonder if some package(s) add immediate functions for some attributes which then leads to both time and memory usage increases

**Comment by hulpke** [comment]
@alex-konovalov 
Is this a test with any assertions turned on? If so, does it also happen if assertions are turned off?

Best,

    Alexander

**Comment by alex-konovalov** [comment]
@hulpke all timings and memory usage in the table above, as well as the break loop backtrace are reproducible with assertions turned off.

**Comment by hulpke** [comment]
@alex-konovalov 
Indeed, thanks. I just tested in my version and loading semigroups increases runtime by a factor two. Puzzling...

**Comment by alex-konovalov** [comment]
@hulpke this is the case when groups are isomorphic. Maybe it takes longer time to find an isomorphism because of different random states/ heuristics / ordering of things etc.

**Comment by hulpke** [comment]
I've now tried a few runs to check impact of immediate methods. (Yes, I know that I am the curmudgeon who has always whined against the concept of immediate methods, so take the following with a grain of salt if you want so.)

After an initial run to ensure memory has been allocated, I get for the Isomorphism calculation in the master branch (no assertions turned on)

plain: 170s
Immediate methods turned off: 80s
after loading semigroups, immediate methods on: 290s
semigroups, immediate methods off: 265s

So there are two issues: Immediate methods in the main library, and ordinary methods in semigroups.

While it is entirely possible that my code is badly written, Is that what we want immediate methods to do?

Would it make sense to also have a test run with immediate methods turned off to spot such performance hits?

(Caveat: I have a patch that speeds the whole calculation up far more, so one could have this issue go away. )

Testing with `TraceImmediateMethods` brings up the following immediate methods
#I  immediate: IsCommutative
#I  immediate: IsCyclic
#I  immediate: IsFinitelyGeneratedGroup
#I  immediate: IsEmpty
#I  immediate: IsTrivial
#I  immediate: IsNonTrivial
#I  immediate: IsPerfectGroup
#I  immediate: IsSolvableGroup

Some of these are old, some of them have been added just recently. Some likely could be replaces by `TrueMethods`, some by simply setting certain flags when creating groups. (None strikes me as particularly useful for generic calculations, a few actually look a bit ``show-off'')

While they are short, part of the cost is the iterated type change, as a few basic properties are deduced one-by-one.

#I  immediate: Size

would set Size to infinity once IsFinite is set to false. Is this really so valuable?

The semigroups package seems to be involved only with
#I  immediate: IsSemigroupIdeal

Should we make these immediate methods another issue?

**Comment by ChrisJefferson** [comment]
I think we should make it an issue. I knew immediate methods could slow things down, but I didn't guess it would be a 2x slowdown.

**Comment by alex-konovalov** [comment]
@hulpke 
> (Caveat: I have a patch that speeds the whole calculation up far more, so one could have this issue go away. )

Thanks - not sure I exactly understand what do you mean by "this issue" - #2377 or just the performance regression. Just in case to point out that to resolve #2377 one needs to fix running out of memory limit.

**Comment by hulpke** [comment]
@alex-konovalov 
It resolves time and memory limit (but does not revolve the issues of immediate methods and the semigroups package, so I hold off with it).
This improvement has nothing to do with the issues observed here, but is simply using the fact that the automorphism group is in fact solvable.

**Comment by fingolfin** [comment]
I agree that we should look into (a) improving existing immediate methods, if poisble (i.e. make them faster, (b) reducing the number of them, (c) making the code which handles immediate methods faster.

I have several ideas for that. E.g. for immediate methods on properties, in most (all) cases in the library, `InstallTrueMethod` is not quite enough right now. However, this can be changed by  implemented extended true implications, and also `InstallFalseMethod`, as discussed in issue #235. I had started work on that recently for other reasons, but it might be beneficial here, too. (Of course it should be carefully benchmarked, to see if it actually helps, and does not hurt existing cases).

And yeah, we should measure performance regressions more closely (see also issues #260 and #636). In fact I know that @alex-konovalov used to do that, but it's unfortunate not quite easy to setup and maintain (and we *cannot* use Travis for that, as its runtimes fluctuate very heavily; we need to use dedicated hardware; which exists, in St Andres, but AFAIK the results are currently not visibly unless you can VPN to the St Andrews network... We discussed ways to change this in the past, but so far nothing happened (as usual, it's a matter of having the time to work on in it). I dream of having something like http://speed.pypy.org for GAP, but right now we are pretty far away from that.

**Comment by fingolfin** [comment]
Some more notes: If I load `Semigroups`, then many more if its immediate methods beyond `IsSemigroupIdeal` are triggered in my tests. 

I think in the tests, it is also instructive to see what happens for `gap -A`: at least for me, the following test requires 288 seconds with `-A`, but only 135 seconds with the default set of packages.
```gap
G:=PcGroupCode( 741231213963541373679312045151639276850536621925972119311,11664);;
IsomorphismGroups(G,PcGroupCode(CodePcGroup(G),Size(G)))=fail; time;
```

**Comment by fingolfin** [comment]
@alex-konovalov you call this a regression, but I don't understand why. This test was not present in GAP 4.8, so it cannot have regressed "as a test". The only thing that regressed in this sense is the overall testsuite, in the sense that it now may not pass resp. take too long on some configurations.

But the examples given here themselves do not seem to have regressed. To the contrary: In some quick tests I just run, performance seems to be *far* worse for them in GAP 4.8, worse than the worst I've seen in `master` resp. `stable-4.9`. And with that I mean that instead of running for 2-4 minutes on my test machine, they run longer than 15 minutes at least (they have not yet completed).

**Comment by james-d-mitchell** [comment]
I'm slightly puzzled by your comment @fingolfin that "many more of its immediate methods beyond IsSemigroupIdeal are triggered in my tests". In `Semigroups` there are only 16 immediate methods installed, and only 5 of them could possibly triggered by objects not created within the `Semigroups` package (i.e. unless your tests use some objects defined in the `Semigroups` package, then I don't see how any of the other methods could be triggered). 

The immediate method for `IsSemigroupIdeal` is just the following:

~~~~
InstallImmediateMethod(IsSemigroupIdeal, IsSemigroup, 0, IsMagmaIdeal);
~~~~

the many more other 4 are:

~~~~
InstallImmediateMethod(IsTrivial, IsMonoid and HasGeneratorsOfMonoid,
InstallImmediateMethod(IsTrivial, IsMonoid and HasGeneratorsOfSemigroup,
InstallImmediateMethod(IsMonogenicSemigroup,
InstallImmediateMethod(IsMonogenicMonoid,
~~~~

If it is somehow desirable to remove/improve these methods, then please file an issue on the Semigroups package issue tracker, and I'll be happy to resolve them.

**Comment by fingolfin** [comment]
@james-d-mitchell My apologies, I misread a swarming output of `TraceImmediateMethods` in a longer session. Let me try to clarify: Right now, all I see are two immediates from semigroups `IsSemigroupIdeal` and `IndecomposableElements` (see below). Also note: This is not a about "blaming" semigroups, or asking you for changes, but rather we are still trying to figure out what's going on; and it's quite noticeable that having Semigroups loaded has a major impact on speed and memory usage.

Here is an example log:
```
gap> LoadPackage("semigroups",false);
true
gap> TraceImmediateMethods();
gap> G:=Group( (1,2), (1,2,3) );;
#I RunImmediateMethods
#I  immediate: Size at /Users/mhorn/Projekte/GAP/gap.github/lib/coll.gi:174
#I  immediate: IsCyclic at /Users/mhorn/Projekte/GAP/gap.github/lib/grp.gi:34
#I  immediate: IsCommutative at /Users/mhorn/Projekte/GAP/gap.github/lib/magma.gi:190
#I  immediate: IsTrivial at /Users/mhorn/Projekte/GAP/gap.github/lib/magma.gi:124
#I  immediate: IsSemigroupIdeal via ((IsMagma and IsLeftActedOnBySuperset) and (IsMagma and IsRightActedOnBySuperset))
#I  immediate: IndecomposableElements at /Users/mhorn/Projekte/GAP/repos/pkg/pkg/Semigroups/gap/attributes/attr.gi:948
gap> G:=PcGroupCode( 741231213963541373679312045151639276850536621925972119311,11664);;
#I RunImmediateMethods
#I  immediate: IsCyclic at GAPROOT/lib/grp.gi:34
#I  immediate: IsCommutative at GAPROOT/lib/magma.gi:190
#I  immediate: IsTrivial at GAPROOT/lib/magma.gi:124
#I  immediate: IsSemigroupIdeal via ((IsMagma and IsLeftActedOnBySuperset) and (IsMagma and IsRightActedOnBySuperset))
#I  immediate: IndecomposableElements at GAPROOT/pkg/Semigroups/gap/attributes/attr.gi:948
#I RunImmediateMethods
#I  immediate: IsNonTrivial at GAPROOT/lib/coll.gi:134
#I RunImmediateMethods
#I RunImmediateMethods
#I RunImmediateMethods
#I  immediate: IsSemigroupIdeal via ((IsMagma and IsLeftActedOnBySuperset) and (IsMagma and IsRightActedOnBySuperset))
#I  immediate: IndecomposableElements at GAPROOT/pkg/Semigroups/gap/attributes/attr.gi:948
#I RunImmediateMethods
#I  immediate: IsCyclic at GAPROOT/lib/grp.gi:34
#I  immediate: IsFinitelyGeneratedGroup at GAPROOT/lib/grp.gi:22
#I  immediate: IsCommutative at GAPROOT/lib/magma.gi:190
#I  immediate: IsTrivial at GAPROOT/lib/magma.gi:124
#I RunImmediateMethods
#I RunImmediateMethods
#I RunImmediateMethods
#I  immediate: Size at GAPROOT/lib/coll.gi:174
#I  immediate: IsCyclic at GAPROOT/lib/grp.gi:34
#I  immediate: IsCommutative at GAPROOT/lib/magma.gi:190
#I  immediate: IsTrivial at GAPROOT/lib/magma.gi:124
#I  immediate: IsSemigroupIdeal via ((IsMagma and IsLeftActedOnBySuperset) and (IsMagma and IsRightActedOnBySuperset))
#I  immediate: IndecomposableElements at GAPROOT/pkg/Semigroups/gap/attributes/attr.gi:948
#I RunImmediateMethods
#I  immediate: IsTrivial at GAPROOT/lib/coll.gi:116
#I  immediate: IsEmpty at GAPROOT/lib/coll.gi:97
#I  immediate: IsPerfectGroup via IsTrivial
#I  immediate: IsNonTrivial at GAPROOT/lib/coll.gi:134
```


I think a major culprit here is `IndecomposableElements`: it adds to every single group object an attribute, set to `[]`. That adds up quickly: Each empty list takes up 24 bytes of heap storage, pus one master pointer; then the attribute has to be stored, which costs extra storage. So for example, `MemoryUsage(Group( (1,2), (1,2,3) ));` is 178 without semigroups, and 228 with it. That's "just" 50 bytes (in other examples, I see 64, it depends on which attributes are set otherwise). But it adds up, both in time and memory usage, if you compute with millions of groups, as is the case here.

That said, there must be other factors at work here; as I said, we are still exploring.

**Comment by james-d-mitchell** [comment]
@fingolfin Thanks for clearing that up. Just to note further that `IndecomposableElements` does not yet appear in any released version of Semigroups (it's in the master branch for inclusion in version 3.1.0). I'm not completely sure why there is an immediate method for this, perhaps @wilfwilson (as the author of the method) can comment further?

**Comment by wilfwilson** [comment]
I do not remember why I made the immediate method. I'll make a PR to Semigroups for removing it.

**Comment by alex-konovalov** [comment]
@fingolfin I call it a [regression](https://en.wikipedia.org/wiki/Software_regression) because `make teststandard` used to pass in all configurations in stable-4.9 branch (i.e. for GAP 4.9.1 release candidate) and does not pass any more.

**Comment by fingolfin** [comment]
@alex-konovalov if that's how you define it, then moving this *new* test from `teststandard` to `testextra` (or deleting it entirely) is the appropriate fix for `stable-4.9`.

**Comment by alex-konovalov** [comment]
@fingolfin agree without hesitations - although not to testextra (those tests are performed as part of `make teststandard`) but e.g. to `benchmark` directory.

**Comment by alex-konovalov** [comment]
P.S. @fingolfin, should the PR for removing this example from teststandard go into master or stable-4.9?

**Comment by james-d-mitchell** [comment]
Just to mention that the immediate method for `IndecomposableElements` is now removed from the `Semigroups` package master branch, if any of the other immediate methods are causing any issues, then please let us know.

**Comment by alex-konovalov** [comment]
This now happens only in the master branch. In stable-4.9 the test causing the problem has been removed in #2407.

**Comment by fingolfin** [comment]
This should be fixed now, due to PR #2383 and #2387 (resp. #2522)
